Detailed status assertions (#61)
peppelinux/draft-demarco-oauth-status-assertions: draft-demarco-oauth-status-assertions.md

to holders, who can present these to verifier along
with the corresponding digital credentials.
The approach outlined in this document
makes the verifier able to check the status,

> such as the nonb-revocation, of a digital credential

Reviewer: ```suggestion
such as the non-revocation, of a digital credential
```

- MUST enable the offline use cases by employing validation using
a cryptographic signature and the cryptographic public key of the
Credential Issuer.

> - SHOULD NOT contain personal information about the User who owns

Author: ```suggestion
- MUST NOT contain personal information about the User who owns
```

Reviewer 2: I suggest something along the lines of "MUST NOT contain information about the data subject (user) that isn't already made available to the validator"

Author: hmm, thats undermining one of my use cases.

Lets imagine a truck that can be loaded with different products, and each product has a different weight.

The truck is empty waiting to be loaded, it is loaded with copper ore, it drives over a weighing station which creates a status assertion for the truck that asserts: copper, weight.

The rest of the truck credential may have an "unloaded weight", but it won't have "copper".

Let me give a personal credentials use case that centers around new PII disclosures as well.

Lets say a person has no children when their credential is issued, but a few days later then have twins.

The status assertion could be: children: true, or children: 2 ... in either case, it would be new information to a verifier.

If your point is that "determining what is PII is a can of worms"... I agree.

I might recommend we simple avoid the discussion of PII outside of security considerations, and treat the status assertion as conveying information which the issuer can only provide about the subject after the credential has been issued and bound to the subjects public key... this information could be related to time or location or changes in physical traits, the subject of a credential is not necessarily a natural or legal person... see security considerations section __ for guidance on informed consent for personal information related to natural persons, etc....

> - SHOULD NOT contain personal information about the User who owns

Reviewer: ```suggestion
- SHOULD NOT contain personal information about the User, that isn't already made available to the Credential Verifier, who owns
```

> the Digital Credential to which the Status Assertion refers.

Reviewer 2: what I am really looking for is a clear resolution to #31 here by aligning the language as best we can with that of the regulation

> the Digital Credential to which the Status Assertion refers.

Author: Thanks for the pointer.

Let's unpack the requirements:

> not allow providers of electronic attestations of attributes or any other party, after the issuance of the attestation of attributes, to obtain data that allows transactions or user behaviour to be tracked, linked or correlated, or knowledge of transactions or user behaviour to be otherwise obtained, unless explicitly authorised by the user;

The key part of this is last bit.

If the user (holder) provides additional information to the issuer, the assertion can contain new information, including potential tracking.

In the case the holder discloses a status assertion or potentially even as fresh disclosure of a disclosable status assertion.. the holder controls what information other parties receive.

In status assertion the biggest privacy risk comes from issuer tracking, based on when fresh assertions are requested.

Justin mentioned this at the mic line in the oauth session at IETF 120.

We should provide guidance on the privacy trade offs.

> the Digital Credential to which the Status Assertion refers.

Author: I've revised the suggestion to match https://github.com/peppelinux/draft-demarco-oauth-status-assertions/pull/61/files#r1681325358

| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be greater than the value set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound. The value SHOULD be set to `sha-256`. | this specification |

> | **credential_status_validity**| Boolean value determining the absolute validity of the Credential to which the Status Assertion is bound | this specification |

Reviewer: ```suggestion
| **credential_status_validity**| Boolean value indicating the absolute validity of the Credential linked to the Status Assertion. This parameter is REQUIRED, and the Verifier MUST verify its presence and value to assess the Credential's validity. | this specification |
```

> | **credential_status_validity**| Boolean value indicating the absolute validity of the Credential linked to the Status Assertion. This parameter is REQUIRED, and the Verifier MUST verify its presence and value to assess the Credential's validity. | this specification |

Reviewer: `credential_status_detail` missing here

FYI here I propose also `credential_status` -> `credential_status_detail`

@fmarino-ipzs ^

allowing Verifiers to examine and make educated conclusions based on a
variety of scenarios.

# Detailed Status Assertions

Status Assertions can introduce a more accurate level of detail, and therefore not necessarly limited to simple boolean information.
This enables Verifier policies to be conditioned on the presence of secured information, instead of the absence of information.
This section proposes syntax to support detailed assertions.
The `credential_status_validity` claim MUST be present and be either `true` or `false`.
The `credential_status` claim MAY be present and if present MUST be an object.
The semantics of the claims within the `credential_status` object are determined by the Credential Issuer.

An example of a boolean status is:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",
    "credential_status_validity": false,
    "credential_status": {

>       "suspended": true,

Author: ```suggestion
      "revoked": false,
      "suspended": true,
```

allowing Verifiers to examine and make educated conclusions based on a
variety of scenarios.

# Complex Status Assertions

> Status assertions can be complex, and are not limited to simple boolean information.

Reviewer: ```suggestion
Status Assertions can introduce a more accurate level of detail, and therefore not necessarly limited to simple boolean information.
```

> This enables verifier policies to be conditioned on the presence of secured information, instead of the absence of information.

Reviewer: ```suggestion
This enables Verifier policies to be conditioned on the presence of secured information, instead of the absence of information.
```

This section proposes syntax to support complex assertions.

> The `validity` claim MUST be present and be either `true` or `false`.

Author: ```suggestion
The `credential_status_validity` claim MUST be present and be either `true` or `false`.
```

> The `validity_reasons` claim MAY be present and if present MUST be an object.

Author: ```suggestion
The `credential_status` claim MAY be present and if present MUST be an object.
```

> The semantics of the claims within the `validity_reasons` object are determined by the issuer.

Author: ```suggestion
The semantics of the claims within the `credential_status` object are determined by the issuer.
```

> The semantics of the claims within the `credential_status` object are determined by the issuer.

Reviewer: ```suggestion
The semantics of the claims within the `credential_status` object are determined by the Credential Issuer.
```

An example of a boolean status is:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",

>     "validity": false,

Author: ```suggestion
    "credential_status_validity": false,
```

>     "validity_reasons": {

Reviewer: wdyt about

`validity` -> `credential_status_validity`
`validity_reason` -> `credential_status`

Reviewer: where `credential_status_validity` is required, therefore all the other section must be updated accordingly

>     "validity_reasons": {

Author: ```suggestion
    "credential_status": {
```

>       "suspended": true,

Reviewer: What about define at least two known credential_statuses

suspendend, revoked

(we might find them also in the EUDIW ARF 1.4)

    },
    "cnf": {
      "jwk": {

>         "kid": "1wL9OG_AZOniODhs2xvMchvFcBWhpQPz3uW29LhqvI0",

Reviewer: ```suggestion
```

        "alg": "ES256",
        "kty": "EC",
        "crv": "P-256",
        "x": "_2ySUmWFjwmraNlo15r6dIBXerVdy_NpJuwAKJMFdoc",
        "y": "MV3C88MhhEMba6oyMBWuGeB3dKHP4YADJmGyJwwILsk"
      }
    }
}
~~~

An example of an enumeration status is:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",

>     "validity": false,

Author: ```suggestion
    "credential_status_validity": false,
```

>     "validity_reasons": {

Author: ```suggestion
    "credential_status_validity": {
```

      "state": "suspended", // or "revoked"
    },
    "cnf": {
      "jwk": {

>         "kid": "1wL9OG_AZOniODhs2xvMchvFcBWhpQPz3uW29LhqvI0",

Reviewer: ```suggestion
```

        "alg": "ES256",
        "kty": "EC",
        "crv": "P-256",
        "x": "_2ySUmWFjwmraNlo15r6dIBXerVdy_NpJuwAKJMFdoc",
        "y": "MV3C88MhhEMba6oyMBWuGeB3dKHP4YADJmGyJwwILsk"
      }
    }
}
~~~

An example of dynamic status using a small matrix:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",

>     "validity": true,

Author: ```suggestion
    "credential_status_validity": true,
```

>     "validity_reasons": {

Author: ```suggestion
    "credential_status": {
```

      "preferences": [[1, 0.25, 0.76 ...] ...]
    },
    "cnf": {
      "jwk": {

>         "kid": "1wL9OG_AZOniODhs2xvMchvFcBWhpQPz3uW29LhqvI0",

Reviewer: ```suggestion
```

allowing Verifiers to examine and make educated conclusions based on a
variety of scenarios.

# Detailed Status Assertions

Status Assertions can introduce a more accurate level of detail, and therefore not necessarly limited to simple boolean information.
This enables Verifier policies to be conditioned on the presence of secured information, instead of the absence of information.
This section proposes syntax to support detailed assertions.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 2: Pending beeing neither valid or invalid...

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer: pending is an half state, my pythonic mind says that 0.5 is not true

![immagine](https://github.com/user-attachments/assets/d5bacd18-7e4b-4511-8615-c4d9be462d34)

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Author: In the W3C Bitstring Status list, there is a purpose called "message" which is not revoked or suspended... You can use it to signal any message the issuer chooses.... Including pending.

The more I have thought about this, the more I conclude that credential status is a separate category of validity information... It's not as simple as being inside or outside a validity period... Or revoked or suspended being invalid and their opposites being valid.

I think it's better to allow the verifier to determine validity, and for the issuer to simply focus on clearly communicating.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 2: so why don't we elliminate the boolean entirely and just have an IANA registry for the status semantic - the spec would register "valid" and "invalid" as two standard semantics. Everybody else get to define their own. The problem with mixing a boolean with "extra information" is that you get into a criciality issue: what if somebody defines a semantic that critically modifies the truthyness of the boolean. This is something that has plagued extensions in X509 btw.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Author: I have come to the position that boolean values are a mistake for status assertions... and that it is verifier policy that determines if a status is cause to reject a presentation.

I've also come to the conclusion that IANA registries are not necessarily the best scalable solution for status assertions in complex domains (chemistry, logistics, healthcare, cyber).

If there are assertions that need to be "globally understood" then IANA makes sense, but I don't see a global registry that starts with "suspended" or "revoked" growing well.

I think many of the entries that might be entered in such a registry are probably much more domain specific.

The guidance to experts would typically be, "make sure there is no existing registry entry that solves this problem"... I imagine adjudicating new requests for all use cases in a single registry to be extremely difficult... and producing the same outcome you mention above, just with multiple status options, or multiple status values... criticality is lost, and verifier behavior becomes undefined.

In my mental model, status assertions are always credential type specific.

Although it might look like both a business license and a drivers license are "revoked"... "revoked" actually means different things, depending on the credential type.

We want to ensure that "revoked" does not mean different things for the same credential type across different issuers.

This is better accomplished with credential type registries, than status registries.

In a credential type registry, you could have:

+------------------+------------------------+
|  type                       |  assertions                       |
+------------------+------------------------+
|  drivers license      |  suspended, revoked      |
+------------------+------------------------+
|  business license   |  pending, revoked           |
+------------------+------------------------+

Statuses are relative to a credential type, and complexity and interoperability are constrained by "registered credential types"... In order to understand "drivers license", you MUST understand "suspended" or "revoked".

This approach also works in the open world model. 

For example, we use JSON Schema to enable this kind of status scheme for W3C Verifiable Credentials in a supply chain context.

The issuer can publish the documentation that verifiers can use to determine how to process the credential, and the status assertions for each credential type that the issuer supports.

If the credential contains a reference to its own schema, there is a single point point of discovery for all this information.

Coming back to the spec text we have.

I think it is fine for the issuer to try to be helpful and tell the verifier "new information, don't reject this", or "new information, please reject this".

But its still going to be the verifier's decision what to do... In that sense, I don't see the value of the boolean as being particularly high.

If the credential type tells you there must be a status assertion and the verifier does not get one, reject.
Otherwise, let the verifier apply their policy to the credential and the status assertion, and determine if they should reject or accept.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 3: In StatusList we have come to the same conclusion as Leif and we already have a regristration for the IANA Status semantic, you can think of piggybacking on this one. This is a potential for symmtry and symbioses between the specs.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 2: That is not a bad idea

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Author: I'm very supportive of aligning the status list and status assertion claims and semantics everywhere we can.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Author: I filed https://github.com/peppelinux/draft-demarco-oauth-status-assertions/issues/81

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer: I support aligning the specs terms, since status assertions was born as a status list feature/extension

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 4: @paulbastian could you please provide reference for the StatusList IANA Status semantics registration.

> The `credential_status_validity` claim MUST be present and be either `true` or `false`.

Reviewer 3: I refer to this section: https://drafts.oauth.net/draft-ietf-oauth-status-list/draft-ietf-oauth-status-list.html#name-status-types
I just observed though that we are missing a corresponding subsection in the IANA as appendix, which I will fix shortly.

The `credential_status` claim MAY be present and if present MUST be an object.
The semantics of the claims within the `credential_status` object are determined by the Credential Issuer.

An example of a boolean status is:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",
    "credential_status_validity": false,

>     "credential_status": {

Reviewer: ```suggestion
    "credential_status_detail": {
```

      "revoked": false,
      "suspended": true,
    },
    "cnf": {
      "jwk": {
        "alg": "ES256",
        "kty": "EC",
        "crv": "P-256",
        "x": "_2ySUmWFjwmraNlo15r6dIBXerVdy_NpJuwAKJMFdoc",
        "y": "MV3C88MhhEMba6oyMBWuGeB3dKHP4YADJmGyJwwILsk"
      }
    }
}
~~~

An example of an enumeration status is:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",
    "credential_status_validity": false,

>     "credential_status": {

Reviewer: ```suggestion
    "credential_status_detail": {
```

      "state": "suspended", // or "revoked"
    },
    "cnf": {
      "jwk": {
        "alg": "ES256",
        "kty": "EC",
        "crv": "P-256",
        "x": "_2ySUmWFjwmraNlo15r6dIBXerVdy_NpJuwAKJMFdoc",
        "y": "MV3C88MhhEMba6oyMBWuGeB3dKHP4YADJmGyJwwILsk"
      }
    }
}
~~~

An example of dynamic status using a small matrix:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion+jwt",
    "kid": "w8ZOZRcx21Zpry7H-0VLBsH7Wf7WXb6TeK3qVMCpY44"
}
.
{
    "iss": "https://issuer.example.org",
    "iat": 1504699136,
    "exp": 1504785536,
    "credential_hash": "xnlAq6Ma8fgu1z4hdGphJnKLulaVHpLCFeZFUGpQ2dA",
    "credential_hash_alg": "sha-256",
    "credential_status_validity": true,

>     "credential_status": {

Reviewer: ```suggestion
    "credential_status_detail": {
```

a deliberate effort to balance security and privacy needs in the
Digital Credential ecosystem.

## Validity Reasons

> Status assertions can reveal information about the holder or subject that was not commited to in the original credential issuance.

Reviewer: ```suggestion
Status Assertions may disclose details about the Holder or subject that were not initially committed to during the original Credential issuance. This can potentially expose additional information that was not part of the original credentialing process.
```

ok, we need to align the general requirements mandating not having subject's information within the assertions

a deliberate effort to balance security and privacy needs in the
Digital Credential ecosystem.

## Validity Reasons

> Status Assertions may disclose details about the Holder or subject that were not initially committed to during the original Credential issuance. This can potentially expose additional information that was not part of the original credentialing process.

Reviewer: ```suggestion
Depending by the scopes of how the detailed Status Assertions are implemented, these may disclose details about the Holder or subject that were not initially committed to during the original Credential issuance. This can potentially expose additional information that was not part of the original credentialing process.
```
